### PR TITLE
BUG: reading windows utf8 filenames in py3.6

### LINKED ISSUE #15086: OSError when reading file with accents in file path
#### Code Sample, a copy-pastable example if possible
`test.txt` and `test_é.txt` are the same file, only the name change:
```python
pd.read_csv('test.txt')
Out[3]: 
   1 1 1
0  1 1 1
1  1 1 1

pd.read_csv('test_é.txt')
Traceback (most recent call last):

  File "<ipython-input-4-fd67679d1d17>", line 1, in <module>
    pd.read_csv('test_é.txt')

  File "d:\app\python36\lib\site-packages\pandas\io\parsers.py", line 646, in parser_f
    return _read(filepath_or_buffer, kwds)

  File "d:\app\python36\lib\site-packages\pandas\io\parsers.py", line 389, in _read
    parser = TextFileReader(filepath_or_buffer, **kwds)

  File "d:\app\python36\lib\site-packages\pandas\io\parsers.py", line 730, in __init__
    self._make_engine(self.engine)

  File "d:\app\python36\lib\site-packages\pandas\io\parsers.py", line 923, in _make_engine
    self._engine = CParserWrapper(self.f, **self.options)

  File "d:\app\python36\lib\site-packages\pandas\io\parsers.py", line 1390, in __init__
    self._reader = _parser.TextReader(src, **kwds)

  File "pandas\parser.pyx", line 373, in pandas.parser.TextReader.__cinit__ (pandas\parser.c:4184)

  File "pandas\parser.pyx", line 669, in pandas.parser.TextReader._setup_parser_source (pandas\parser.c:8471)

OSError: Initializing from file failed

```
#### Problem description

Pandas return OSError when trying to read a file with accents in file path.

The problem is new (Since I upgraded to Python 3.6 and Pandas 0.19.2)

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.0.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 94 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: fr
LOCALE: None.None

pandas: 0.19.2
nose: None
pip: 9.0.1
setuptools: 32.3.1
Cython: 0.25.2
numpy: 1.11.3
scipy: 0.18.1
statsmodels: None
xarray: None
IPython: 5.1.0
sphinx: 1.5.1
patsy: None
dateutil: 2.6.0
pytz: 2016.10
blosc: None
bottleneck: 1.2.0
tables: None
numexpr: 2.6.1
matplotlib: 1.5.3
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.999999999
httplib2: None
apiclient: None
sqlalchemy: 1.1.4
pymysql: None
psycopg2: None
jinja2: 2.9.3
boto: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by vnlitvinov**
- [x] closes #15086 properly
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Test that was added in prior fix was failing on my machine, where my locale was set to `CP-1251` (which apparently cannot handle some unicode symbols in the test even when using MBCS, as MBCS encoding is locale-dependent).

This PR fixes the issue removing the workaround by switching to converting UTF8 to UTF16 and using Win32 Unicode file API (much like Python >= 3.6 does internally).

P.S. This as a bonus adds possibility to use `\\?\C\long\file\name` addressing if desired to overcome 260 symbols limitation on path length.

### CODE REVIEW

**Comment by simonjayhawkins** [comment]
> Test that was added in prior fix was failing on my machine, where my locale was set to `CP-1251`

is there an open issue about this?

cc @gfyoung

**Comment by vnlitvin** [comment]
I didn't open an issue when I faced that, instead I have fixed the code.
Should I file an issue before the code could be accepted?

**Comment by simonjayhawkins** [comment]
It normally helps to create a code sample which can be used as the test and for others to recreate the problem.

**Comment by vnlitvin** [comment]
The test that failed was the same test that was introduced in #24758, it just didn't pass on my machine, most likely due to different locale set up. My fix is locale-independent, so it should pass on any machine with any locale set.

**Comment by simonjayhawkins** [comment]
so i guess that `test_filename_with_special_chars` should be parametrised over different locales. would monkeypatch allow this here?

**Comment by vnlitvin** [comment]
My diggings show that when Python faces a request to convert from Unicode to bytestring using `mbcs` encoding on Windows it uses `CP_ACP` as codepage for `WideCharToMultiByte` argument (that is, default ANSI code page).

According to https://stackoverflow.com/a/9039278 there is no programmatic way to change `CP_ACP` value, as this is a system-wide setting controlled via Region settings in Windows Control Panel.

So, in short, the answer to this question:


> would monkeypatch allow this here?

is "no" :(

What _does_ seem to be possible is to change the test so it would fail _both_ for `1252` and `1251` codepages unless fixed as I do here by using a superset of characters some of which are not representable in 1251 and some in 1252.

**Comment by gfyoung** [comment]
> is there an open issue about this?

Not that I'm aware of.

> What does seem to be possible is to change the test so it would fail both for 1252 and 1251 codepages unless fixed as I do here by using a superset of characters some of which are not representable in 1251 and some in 1252.

I would just add a new test actually.

Second, if you could post the output that you were getting prior to this fix for that test, that would be good for everyone's reference.

Finally, if you could add a whatsnew note for `0.25.0`.  Even though it's fixing an already-closed issue, you're adding a more robust fix.  Otherwise, so far so good!

**Comment by vnlitvin** [comment]
> I would just add a new test actually.

I'm not sure what second test would add here, what I'm thinking is simply crafting a string that isn't representable in either of `1251` or `1252` (my current stab at it is `sé-es-vé-sй.csv` - it cannot be encoded neither as 1251 nor as 1252 bytestrings). The only quirk that remains there is that I'd probably have to disable the test for Python < 3.6 (otherwise it would forever fail there on strange locales - as prior to 3.6 I cannot be sure that what this function gets as an input parameter is a UTF-8-encoded string).

> Second, if you could post the output that you were getting prior to this fix for that test, that would be good for everyone's reference.

Failure without my patch is easy:

```
________________ test_filename_with_special_chars[c_high] ____________________________________

all_parsers = <pandas.tests.io.parser.conftest.CParserHighMemory object at 0x000002CA11F92320>

    def test_filename_with_special_chars(all_parsers):
        # see gh-15086.
        parser = all_parsers
        df = DataFrame({"a": [1, 2, 3]})

        with tm.ensure_clean("sé-es-vé.csv") as path:
            df.to_csv(path, index=False)

>           result = parser.read_csv(path)

pandas\tests\io\parser\test_common.py:1915:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pandas\tests\io\parser\conftest.py:22: in read_csv
    return read_csv(*args, **kwargs)
pandas\io\parsers.py:702: in parser_f
    return _read(filepath_or_buffer, kwds)
pandas\io\parsers.py:429: in _read
    parser = TextFileReader(filepath_or_buffer, **kwds)
pandas\io\parsers.py:895: in __init__
    self._make_engine(self.engine)
pandas\io\parsers.py:1122: in _make_engine
    self._engine = CParserWrapper(self.f, **self.options)
pandas\io\parsers.py:1853: in __init__
    self._reader = parsers.TextReader(src, **kwds)
pandas\_libs\parsers.pyx:388: in pandas._libs.parsers.TextReader.__cinit__
    self._setup_parser_source(source)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   source = source.encode(encoding)
E   UnicodeEncodeError: 'mbcs' codec can't encode characters in position 0--1: invalid character

pandas\_libs\parsers.pyx:687: UnicodeEncodeError
```

> Finally, if you could add a whatsnew note for `0.25.0`

Will do tomorrow.

**Comment by gfyoung** [comment]
> The only quirk that remains there is that I'd probably have to disable the test for Python < 3.6 (otherwise it would forever fail there on strange locales - as prior to 3.6 I cannot be sure that what this function gets as an input parameter is a UTF-8-encoded string)

No worries.  Feel free to skip the test for Python < 3.6 (we're already in talks about phasing out 3.5 support once we finish with 2.7).

**Comment by jreback** [comment]
pls don’t skip for 3.5

**Comment by gfyoung** [comment]
> pls don’t skip for 3.5

@jreback : @vnlitvin fix relies on leveraging Python 3.6+ internal functionality that can't be replicated in Python 3.5.  That is why I suggested he add a new test (instead of modifying the existing one) that can then be skipped.

**Comment by jreback** [comment]
ah I c. carry on then.

**Comment by vnlitvin** [comment]
> the existing one

The problem with existing one is that it would fail on Python 3.5 with e.g. 1251 (Cyrillic) locale set.
What can I add is I can skip the test on non-1252 locales for Python < 3.6 (though it seems that the test is kinda useless for < 3.6 - it basically tests that Windows `open()` function works - doesn't seem to be a job for a Pandas test).

**Comment by gfyoung** [comment]
> What can I add is I can skip the test on non-1252 locales for Python < 3.6

Yep, that sounds like a plan.  Go for it!

**Comment by vnlitvin** [comment]
> [Line 1907:1](https://github.com/pandas-dev/pandas/blob/cf2e0622cb118cabfed709e9e3a39c9b0da26f16/pandas/tests/io/parser/test_common.py#L1907): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

Fixed

File: doc/source/whatsnew/v0.25.0.rst [hunk 1]
```diff
@@ -141,6 +141,7 @@ Bug Fixes
 - Bug in :func:`to_datetime` which would raise an (incorrect) ``ValueError`` when called with a date far into the future and the ``format`` argument specified instead of raising ``OutOfBoundsDatetime`` (:issue:`23830`)
 - Bug in an error message in :meth:`DataFrame.plot`. Improved the error message if non-numerics are passed to :meth:`DataFrame.plot` (:issue:`25481`)
 - Bug in error messages in :meth:`DataFrame.corr` and :meth:`Series.corr`. Added the possibility of using a callable. (:issue:`25729`)
+- Bug in `TextReader()` not properly working with UTF8 on Windows on Python 3.6+ (fix for :issue:`15086` instead of a workaround)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: move to I/O and replace this with 

Bug in ``:func:`read_csv` `` not properly interpreting the UTF8 encoded filenames on Windows on Python 3.6+ (:issue:`15086` )

File: pandas/tests/io/parser/test_common.py [hunk 2]
```diff
@@ -1904,12 +1905,21 @@ def test_suppress_error_output(all_parsers, capsys):
     assert captured.err == ""
 
 
+def __should_skip_utf8_test():
+    if compat.is_platform_windows():
+        import ctypes
```
Comment (jreback) [inline pandas/tests/io/parser/test_common.py]: import at the top

File: pandas/tests/io/parser/test_common.py [hunk 3]
```diff
@@ -1904,12 +1905,21 @@ def test_suppress_error_output(all_parsers, capsys):
     assert captured.err == ""
 
 
+def __should_skip_utf8_test():
```
Comment (jreback) [inline pandas/tests/io/parser/test_common.py]: be more specific here on the name of the skip

File: pandas/tests/io/parser/test_common.py [hunk 4]
```diff
@@ -1904,12 +1905,21 @@ def test_suppress_error_output(all_parsers, capsys):
...
+    if compat.is_platform_windows():
+        import ctypes
+        ansi_codepage = ctypes.cdll.kernel32.GetACP()
+        return ansi_codepage != 1252 and sys.version_info < (3, 6)
```
Comment (jreback) [inline pandas/tests/io/parser/test_common.py]: use PY36

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.25.0.rst (hunk 1 above)
Comment (vnlitvin) [reply doc/source/whatsnew/v0.25.0.rst]: Fixed

File: pandas/tests/io/parser/test_common.py (hunk 2 above)
Comment (vnlitvin) [reply pandas/tests/io/parser/test_common.py]: Fixed

File: pandas/tests/io/parser/test_common.py (hunk 3 above)
Comment (vnlitvin) [reply pandas/tests/io/parser/test_common.py]: Renamed to `__windows_ansi_encoding_not_cp1252`

File: pandas/tests/io/parser/test_common.py (hunk 4 above)
Comment (vnlitvin) [reply pandas/tests/io/parser/test_common.py]: Fixed

**Comment by vnlitvin** [comment]
Note for reviewers: it seems that testing is somehow broken in the upstream, as I didn't change anything related to `numpy` or `pytables`, but they now fail. It seems that the image used for testing has changed.

**Comment by vnlitvin** [comment]
@jreback should I mark as resolved the conversations I fixed?

**Comment by jreback** [comment]
@vnlitvin yes you can resolve things.

we already know that the CI is failing. I will ask you to rebase at some point.

File: pandas/tests/io/parser/test_common.py [hunk 5]
```diff
@@ -1904,12 +1905,20 @@ def test_suppress_error_output(all_parsers, capsys):
...
+
+@pytest.mark.skipif(__windows_ansi_encoding_not_cp1252(),
+                    reason="On Python < 3.6 won't pass on non-1252 codepage")
 def test_filename_with_special_chars(all_parsers):
     # see gh-15086.
     parser = all_parsers
     df = DataFrame({"a": [1, 2, 3]})
 
-    with tm.ensure_clean("sé-es-vé.csv") as path:
+    with tm.ensure_clean("sé-es-vé-sй.csv") as path:
```
Comment (gfyoung) [inline pandas/tests/io/parser/test_common.py]: Don't change the test.  Just add one.

**Comment by jreback** [comment]
can you merge master

File: pandas/tests/io/parser/test_common.py (hunk 5 above)
Comment (vnlitvin) [reply pandas/tests/io/parser/test_common.py]: Fixed

**Comment by vnlitvin** [comment]
> can you merge master

I've rebased on latest master, so commit history will be clean.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @vnlitvin
